File: kattis/ecoins.cpp
#include <bits/stdc++.h>

#define int long long
using namespace std;

signed main() {
    int t; cin >> t;
    while (t--) {
        int m, s; cin >> m >> s;
        vector<pair<int,int>> coins(m, pair<int,int>());
        for (int i = 0; i < m; i++) {
            cin >> coins[i].first >> coins[i].second;
        }

        vector<vector<int>> dp(301, vector<int>(301, 1e9));
        dp[0][0] = 0;
        for (auto [x, y] : coins) {
            for (int i = x; i < dp.size(); i++) {
                for (int j = y; j < dp.size(); j++) {
                    dp[i][j] = min(dp[i][j], dp[i-x][j-y] + 1);
                }
            }
        }
        int ans = 1e9;
        for (int i = 0; i <= s; i++) {
            int target = s * s - i * i;
            int sq = sqrt(target) + 1e-9;
            if (sq * sq != target) continue;
            // cout << sq << endl;
            ans = min(ans, dp[i][sq]);
            ans = min(ans, dp[sq][i]);
        }

        if (ans == 1e9)
            cout << "not possible" << endl;
        else
            cout << ans << endl;

    }
}

File: kattis/ecoins.py
for _ in range(int(input())):
    if _: input()
    m, s = map(int, input().split())
    
    l = []
    for _ in range(m):
        x, y = map(int, input().split())
        l.append((x,y))
    dp = [[float("inf")]*301 for _ in range(301)]

    dp[0][0] = 0
    for x, y in l:
        for i in range(x, len(dp)):
            for j in range(y, len(dp)):
                dp[i][j] = min(dp[i][j], 1 + dp[i - x][j - y])
    
    ans = float("inf")
    for i in range(s+1):
        target = (s ** 2 - i ** 2) ** 0.5
        if abs(target - round(target)) > 1e-9 or round(target) < 0:
            continue
        # print(target)
        ans = min(ans, dp[i][round(target)])


    print(ans if ans != float("inf") else "not possible")